Use "info" field in spec

This is necessary for the CLI. We should use the [`info`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#infoObject) field to get the title, description, version, etc. (only if they exist - I don't think it's necessary to force that in CLI). An edge case would be that passing name and version through CLI args will override whatever name and version we obtained from the spec.

Hi @wafflespeanut, I am attempting to do `state.set_meta` in the `generate` fn,
```rust
if let Some(m) = state.get_meta() {
    let meta = m.borrow_mut();
    if meta.name.is_none() {
         meta.name = Some(api.info.title)
    }
    if meta.version.is_none() {
        meta.version = Some(api.info.version)
    }
    state.set_meta(meta)
 }
```

wondering what am I doing wrong here to get this 🤔 
```
error[E0308]: mismatched types
   --> src/v2/codegen/emitter.rs:223:28
    |
223 |             state.set_meta(meta)
    |                            ^^^^ expected struct `v2::codegen::CrateMeta`, found struct `std::cell::RefMut`
    |
    = note: expected type `v2::codegen::CrateMeta`
               found type `std::cell::RefMut<'_, v2::codegen::CrateMeta, >`
```

Thanks @wafflespeanut. Have another one :) 
```
error[E0596]: cannot borrow `*state` as mutable, as it is behind a `&` reference
...
223 |             state.set_meta(meta.clone())
    |             ^^^^^ `state` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```
I tried this but still no success

```rust
let state = &mut self.state()
```

--- a/core/src/v2/mod.rs
+++ b/core/src/v2/mod.rs
@@ -34,6 +34,7 @@ impl<S: Schema + Default> Api<S> {
         resolver.resolve()?;
         Ok(Api {
             swagger: self.swagger,
+            info: self.info,
             definitions: resolver.defs,
             paths: resolver.paths,
             base_path: self.base_path,

--- a/core/src/v2/models.rs
+++ b/core/src/v2/models.rs
@@ -125,6 +125,7 @@ pub struct GenericApi<S> {
     /// the spec was provided.
     #[serde(skip)]
     pub spec_format: SpecFormat,
+    pub info: Info,
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
@@ -223,6 +224,45 @@ pub enum SchemaRepr<S> {
         new: ArcRwLock<S>,
         old: ArcRwLock<S>,
     },
+}
+
+/// Info Object
+///
+/// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#infoObject
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct Info {
+    pub version: String,
+    pub title: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub contact: Option<Contact>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub license: Option<License>,
+}
+
+/// Contact Object
+///
+/// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#contactObject
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct Contact {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub url: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub email: Option<String>,
+}
+
+/// License Object
+///
+/// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#licenseObject
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct License {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub url: Option<String>,
 }
 
 /// Path item.

--- a/src/v2/codegen/emitter.rs
+++ b/src/v2/codegen/emitter.rs
@@ -1,5 +1,6 @@
 use super::object::{ApiObject, ObjectField, OpRequirement, Parameter, Response};
 use super::state::{ChildModule, EmitterState};
+use super::CrateMeta;
 use crate::error::PaperClipError;
 use crate::v2::{
     im::ArcRwLock,
@@ -13,14 +14,13 @@ use crate::v2::{
 use failure::Error;
 use heck::{CamelCase, SnekCase};
 use itertools::Itertools;
-use url::Host;
-
 use std::collections::{BTreeSet, HashSet};
 use std::fmt::Debug;
 use std::fs;
 use std::ops::Deref;
 use std::path::{Path, PathBuf};
 use std::sync::Arc;
+use url::Host;
 
 /// Identifier used for `Any` generic parameters in struct definitions.
 pub(super) const ANY_GENERIC_PARAMETER: &str = "Any";
@@ -210,6 +210,15 @@ pub trait Emitter: Sized {
     fn generate(&self, api: &Api<Self::Definition>) -> Result<(), Error> {
         let state = self.state();
         state.reset_internal_fields();
+
+        let m = state.get_meta();
+
+        if m.borrow().is_none() {
+            let mut meta = CrateMeta::default();
+            meta.name = Some(api.info.title.clone());
+            meta.version = Some(api.info.version.clone());
+            state.set_meta(meta);
+        }
 
         // Add default coders.
         let mut coders = api.coders.clone();

--- a/src/v2/codegen/state.rs
+++ b/src/v2/codegen/state.rs
@@ -38,7 +38,7 @@ pub struct EmitterState {
     /// If crate metadata is specified, then `lib.rs` and `Cargo.toml` are generated
     /// along with the modules. This is gated behind `"cli"` feature.
     #[cfg(feature = "cli")]
-    crate_meta: Option<Rc<RefCell<CrateMeta>>>,
+    crate_meta: Rc<RefCell<Option<CrateMeta>>>,
 
     /* MARK: Internal fields that should be reset for each session. */
     /// Maps parent mod to immediate children. Used for declaring modules.
@@ -325,9 +325,9 @@ pub mod util {
             return Ok(());
         }
 
-        if let Some(m) = self.infer_crate_meta()? {
+        let cm = self.infer_crate_meta()?;
+        if let Some(meta) = cm.borrow().as_ref() {
             // Clap YAML
-            let meta = m.borrow();
             let clap_yaml = root.with_file_name("app.yaml");
             let base_content = template::render(
                 TEMPLATE::CLAP_YAML,
@@ -362,7 +362,9 @@ pub mod util {
     fn is_cli(&self) -> Result<bool, Error> {
         Ok(self
             .infer_crate_meta()?
-            .map(|m| m.borrow().is_cli)
+            .borrow()
+            .as_ref()
+            .map(|m| m.is_cli)
             .unwrap_or(false))
     }
 }
@@ -372,19 +374,23 @@ pub mod util {
 #[cfg(feature = "cli")]
 impl EmitterState {
     /// Sets the crate metadata for this session.
-    pub fn set_meta(&mut self, meta: CrateMeta) {
-        self.crate_meta = Some(Rc::new(RefCell::new(meta)));
+    pub fn set_meta(&self, meta: CrateMeta) {
+        *self.crate_meta.borrow_mut() = Some(meta)
+    }
+
+    pub(super) fn get_meta(&self) -> Rc<RefCell<Option<CrateMeta>>> {
+        self.crate_meta.clone()
     }
 
     /// Checks whether this session is for emitting a crate.
     fn is_crate(&self) -> bool {
-        self.crate_meta.is_some()
+        self.crate_meta.borrow().is_some()
     }
 
     /// Returns the path to the root module.
     fn root_module_path(&self) -> PathBuf {
-        if let Some(m) = self.crate_meta.as_ref() {
-            let meta = m.borrow();
+        let cm = self.crate_meta.borrow();
+        if let Some(meta) = cm.as_ref() {
             if meta.is_cli {
                 self.working_dir.join("main.rs")
             } else {
@@ -401,12 +407,13 @@ impl EmitterState {
         let is_cli = self.is_cli()?;
         man_path.set_file_name("Cargo.toml");
 
-        let m = match self.infer_crate_meta()? {
+        let cm = self.infer_crate_meta()?;
+        let m = cm.borrow();
+        let meta = match m.as_ref() {
             Some(c) => c,
             None => return Ok(()),
         };
 
-        let meta = m.borrow();
         if self.is_crate() {
             let contents = template::render(
                 TEMPLATE::CARGO_MANIFEST,
@@ -425,9 +432,9 @@ impl EmitterState {
     }
 
     /// Validates crate metadata, sets the unset fields and returns a reference.
-    fn infer_crate_meta(&self) -> Result<Option<Rc<RefCell<CrateMeta>>>, Error> {
-        if let Some(m) = self.crate_meta.as_ref() {
-            let mut meta = m.borrow_mut();
+    fn infer_crate_meta(&self) -> Result<Rc<RefCell<Option<CrateMeta>>>, Error> {
+        let mut cm = self.crate_meta.borrow_mut();
+        if let Some(meta) = &mut *cm {
             if meta.name.is_none() {
                 meta.name = Some(
                     fs::canonicalize(&self.working_dir)?
@@ -462,11 +469,16 @@ impl EmitterState {
 #[cfg(not(feature = "cli"))]
 impl EmitterState {
     /// This is a no-op.
-    pub fn set_meta(&mut self, _: CrateMeta) {}
+    pub fn set_meta(&self, _: CrateMeta) {}
+
+    /// This is a no-op
+    pub fn get_meta(&self) -> Rc<RefCell<Option<CrateMeta>>> {
+        Rc::new(RefCell::new(None))
+    }
 
     /// Always returns `Ok(None)`
-    fn infer_crate_meta(&self) -> Result<Option<Rc<RefCell<CrateMeta>>>, Error> {
-        Ok(None)
+    fn infer_crate_meta(&self) -> Result<Rc<RefCell<Option<CrateMeta>>>, Error> {
+        Ok(Rc::new(RefCell::new(None)))
     }
 
     /// Always returns the path to `mod.rs` in root.
@@ -510,7 +522,7 @@ impl Default for EmitterState {
             mod_prefix: "crate::",
             ns_sep: ".",
             #[cfg(feature = "cli")]
-            crate_meta: None,
+            crate_meta: Rc::new(RefCell::new(None)),
             base_url: RefCell::new("https://example.com".parse().expect("invalid URL?")),
             def_mods: RefCell::new(HashMap::new()),
             rel_paths: RefCell::new(HashSet::new()),

--- a/tests/pet-v2.yaml
+++ b/tests/pet-v2.yaml
@@ -1,4 +1,9 @@
 swagger: "2.0"
+info:
+  version: 1.0.0
+  title: Petstore
+  license:
+    name: MIT
 host: pets.com:8888
 basePath: /api
 definitions:

--- a/tests/test_app.rs
+++ b/tests/test_app.rs
@@ -75,6 +75,7 @@ fn test_simple_app() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {
                     "Pet": {
                       "properties": {
@@ -285,6 +286,7 @@ fn test_params() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {
                     "BadgeBody":{
                       "properties":{
@@ -444,6 +446,7 @@ fn test_map_in_out() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {
                     "Image": {
                       "properties": {
@@ -519,6 +522,7 @@ fn test_list_in_out() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {
                     "Pet": {
                       "properties": {
@@ -624,6 +628,7 @@ fn test_impl_traits() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {
                     "Pet": {
                       "properties": {
@@ -731,6 +736,7 @@ fn test_multiple_method_routes() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {},
                   "paths": {
                     "/foo": {
@@ -816,6 +822,7 @@ fn test_custom_extractor_empty_schema() {
             check_json(
                 &mut resp,
                 json!({
+                  "info":{"title":"","version":""},
                   "definitions": {},
                   "paths": {
                     "/": {

--- a/tests/test_emitter.rs
+++ b/tests/test_emitter.rs
@@ -12,6 +12,9 @@ fn test_templated_path_missing_parameters() {
         b"
 swagger: \"2.0\"
 definitions: {}
+info:
+  title:  \"Petstore\"
+  version: \"1.0.0\"
 paths:
   /pets/{petId}:
     get:
@@ -40,6 +43,9 @@ fn test_templated_path_uniqueness() {
         b"
 swagger: \"2.0\"
 definitions: {}
+info:
+  title:  \"Petstore\"
+  version: \"1.0.0\"
 paths:
   /store/{id1}/pets/{id2}:
     parameters:

